File: packages/text-recognition/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/textrecognition/classes/results/ProcessImageResult.java
package io.capawesome.capacitorjs.plugins.mlkit.textrecognition.classes.results;

import android.graphics.Point;
import android.graphics.Rect;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.google.mlkit.vision.text.Text;
import io.capawesome.capacitorjs.plugins.mlkit.textrecognition.interfaces.Result;
import java.util.List;

public class ProcessImageResult implements Result {

    @NonNull
    private final Text text;

    public ProcessImageResult(@NonNull Text text) {
        this.text = text;
    }

    @Override
    @NonNull
    public JSObject toJSObject() {
        JSObject result = new JSObject();
        result.put("text", text.getText());
        result.put("blocks", createBlocksResult(text.getTextBlocks()));
        return result;
    }

    private JSArray createBlocksResult(@NonNull List<Text.TextBlock> blocks) {
        JSArray result = new JSArray();
        for (Text.TextBlock block : blocks) {
            JSObject blockResult = new JSObject();
            blockResult.put("text", block.getText());
            putBoundingBox(blockResult, block.getBoundingBox());
            putCornerPoints(blockResult, block.getCornerPoints());
            putRecognizedLanguage(blockResult, block.getRecognizedLanguage());
            blockResult.put("lines", createLinesResult(block.getLines()));
            result.put(blockResult);
        }
        return result;
    }

    private JSArray createLinesResult(@NonNull List<Text.Line> lines) {
        JSArray result = new JSArray();
        for (Text.Line line : lines) {
            JSObject lineResult = new JSObject();
            lineResult.put("text", line.getText());
            putBoundingBox(lineResult, line.getBoundingBox());
            putCornerPoints(lineResult, line.getCornerPoints());
            putRecognizedLanguage(lineResult, line.getRecognizedLanguage());
            lineResult.put("elements", createElementsResult(line.getElements()));
            result.put(lineResult);
        }
        return result;
    }

    private JSArray createElementsResult(@NonNull List<Text.Element> elements) {
        JSArray result = new JSArray();
        for (Text.Element element : elements) {
            JSObject elementResult = new JSObject();
            elementResult.put("text", element.getText());
            putBoundingBox(elementResult, element.getBoundingBox());
            putCornerPoints(elementResult, element.getCornerPoints());
            putRecognizedLanguage(elementResult, element.getRecognizedLanguage());
            result.put(elementResult);
        }
        return result;
    }

    private void putBoundingBox(@NonNull JSObject target, @Nullable Rect boundingBox) {
        if (boundingBox == null) {
            return;
        }
        JSObject result = new JSObject();
        result.put("left", boundingBox.left);
        result.put("top", boundingBox.top);
        result.put("right", boundingBox.right);
        result.put("bottom", boundingBox.bottom);
        target.put("boundingBox", result);
    }

    private void putCornerPoints(@NonNull JSObject target, @Nullable Point[] cornerPoints) {
        if (cornerPoints == null) {
            return;
        }
        JSArray result = new JSArray();
        for (Point cornerPoint : cornerPoints) {
            JSObject pointResult = new JSObject();
            pointResult.put("x", cornerPoint.x);
            pointResult.put("y", cornerPoint.y);
            result.put(pointResult);
        }
        target.put("cornerPoints", result);
    }

    private void putRecognizedLanguage(@NonNull JSObject target, @Nullable String recognizedLanguage) {
        if (recognizedLanguage == null || recognizedLanguage.isEmpty()) {
            return;
        }
        target.put("recognizedLanguage", recognizedLanguage);
    }
}

File: packages/text-recognition/ios/Plugin/Classes/Results/ProcessImageResult.swift
import Foundation
import Capacitor
import MLKitTextRecognition

@objc public class ProcessImageResult: NSObject, Result {
    let text: Text

    init(text: Text) {
        self.text = text
    }

    @objc public func toJSObject() -> AnyObject {
        var result = JSObject()
        result["text"] = text.text
        result["blocks"] = createBlocksResult(text.blocks)
        return result as AnyObject
    }

    private func createBlocksResult(_ blocks: [TextBlock]) -> JSArray {
        var result = JSArray()
        for block in blocks {
            var blockResult = JSObject()
            blockResult["text"] = block.text
            blockResult["boundingBox"] = createBoundingBoxResult(block.frame)
            blockResult["cornerPoints"] = createCornerPointsResult(block.cornerPoints)
            if let recognizedLanguage = createRecognizedLanguageResult(block.recognizedLanguages) {
                blockResult["recognizedLanguage"] = recognizedLanguage
            }
            blockResult["lines"] = createLinesResult(block.lines)
            result.append(blockResult)
        }
        return result
    }

    private func createLinesResult(_ lines: [TextLine]) -> JSArray {
        var result = JSArray()
        for line in lines {
            var lineResult = JSObject()
            lineResult["text"] = line.text
            lineResult["boundingBox"] = createBoundingBoxResult(line.frame)
            lineResult["cornerPoints"] = createCornerPointsResult(line.cornerPoints)
            if let recognizedLanguage = createRecognizedLanguageResult(line.recognizedLanguages) {
                lineResult["recognizedLanguage"] = recognizedLanguage
            }
            lineResult["elements"] = createElementsResult(line.elements)
            result.append(lineResult)
        }
        return result
    }

    private func createElementsResult(_ elements: [TextElement]) -> JSArray {
        var result = JSArray()
        for element in elements {
            var elementResult = JSObject()
            elementResult["text"] = element.text
            elementResult["boundingBox"] = createBoundingBoxResult(element.frame)
            elementResult["cornerPoints"] = createCornerPointsResult(element.cornerPoints)
            result.append(elementResult)
        }
        return result
    }

    private func createBoundingBoxResult(_ frame: CGRect) -> JSObject {
        var result = JSObject()
        result["left"] = Int(frame.origin.x)
        result["top"] = Int(frame.origin.y)
        result["right"] = Int(frame.origin.x + frame.size.width)
        result["bottom"] = Int(frame.origin.y + frame.size.height)
        return result
    }

    private func createCornerPointsResult(_ cornerPoints: [NSValue]) -> JSArray {
        var result = JSArray()
        for cornerPoint in cornerPoints {
            let point = cornerPoint.cgPointValue
            var pointResult = JSObject()
            pointResult["x"] = Int(point.x)
            pointResult["y"] = Int(point.y)
            result.append(pointResult)
        }
        return result
    }

    private func createRecognizedLanguageResult(_ recognizedLanguages: [TextRecognizedLanguage]) -> String? {
        for recognizedLanguage in recognizedLanguages {
            if let languageCode = recognizedLanguage.languageCode, !languageCode.isEmpty {
                return languageCode
            }
        }
        return nil
    }
}
